src: build without "using namespace std"

tchaikov opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

ronen-fr commented on src/SimpleRADOSStriper.h:
```diff
@@ -119,8 +119,8 @@ class [[gnu::visibility("default")]] SimpleRADOSStriper
   time last_renewal = time::min();
-  std::chrono::milliseconds lock_keeper_interval = 2000ms;
-  std::chrono::milliseconds lock_keeper_timeout = 30000ms;
+  std::chrono::milliseconds lock_keeper_interval{2000};
```
I'd suggest 2'000, 30'000 etc.
Easier than counting the 0s...

ronen-fr commented on src/ceph_mgr.cc:
```diff
@@ -43,18 +43,18 @@ int main(int argc, const char **argv)
   ceph_pthread_setname(pthread_self(), "ceph-mgr");
 
-  vector<const char*> args;
+  std::vector<const char*> args;
```
(\todo: create a 2'nd args_to_vec() with the following signature:
std::optional<std::vector<const char*>> argv_vec(argc, argv)

Cleaner, and will be able to allocate the correct size for the vector.

tchaikov replied:
this is not part of this change.

ronen-fr commented on src/client/SyntheticClient.h:
```diff
@@ -150,7 +150,7 @@ class SyntheticClient {
     r %= contents.size();
 
-    map<string,struct stat*>::iterator it = contents.begin();
+    std::map<std::string,struct stat*>::iterator it = contents.begin();
```
auto?

tchaikov replied:
there are lots of places like this. i did a search-n-replace in most cases. this would be simpler and less error-prune. please feel free to create a follow-up change to use `auto` when appropriate.

ronen-fr commented on src/common/Continuation.h:
```diff
@@ -122,19 +122,20 @@ class Continuation {
   bool _continue_function(int r, int n) {
-    set<int>::iterator stage_iter = stages_in_flight.find(n);
-    ceph_assert(stage_iter != stages_in_flight.end());
+    std::set<int>::iterator in_flight_iter = stages_in_flight.find(n);
```
auto?

tchaikov commented on src/cls/cas/cls_cas_internal.cc:
```diff
@@ -129,7 +129,8 @@ void chunk_refs_t::decode(ceph::buffer::list::const_iterator& p)
   default:
     throw ceph::buffer::malformed_input(
-      "unrecognized chunk ref encoding type "s + stringify((int)t));
+      std::string("unrecognized chunk ref encoding type ") +
```
https://github.com/ceph/ceph/pull/42742#discussion_r686617222

ronen-fr commented on src/common/obj_bencher.h:
```diff
@@ -64,7 +64,7 @@ typedef std::pair<std::string, std::string> Object;
   bool show_time;
   Formatter *formatter = NULL;
-  ostream *outstream = NULL;
+  std::ostream *outstream = NULL;
```
nullptr (if already touching this line)

tchaikov replied:
i'd minimize the change. please note this changeset touches hundreds of files. i'd prefer to be focused.

ronen-fr commented on src/libcephsqlite.cc:
```diff
@@ -510,7 +510,7 @@ static int Open(sqlite3_vfs *vfs, const char *name, sqlite3_file *file,
   }
   auto path = std::string_view(name);
-  if (path == ":memory:"sv) {
+  if (path == ":memory:") {
```
?

rzarzynski commented on src/rgw/rgw_coroutine.h:
```diff
@@ -174,9 +174,8 @@ struct rgw_spawned_stacks {
-    for (vector<RGWCoroutinesStack *>::iterator iter = source->entries.begin();
-         iter != source->entries.end(); ++iter) {
-      add_pending(*iter);
+    for (auto entry : source->entries) {
```
Extra copy?

tchaikov replied:
no. entry is a pointer. but i s/auto/auto*/ for better readability.

cbodley reviewed (approved):
what a massive effort, thanks @tchaikov! approving the rgw changes

ronen-fr commented on src/rgw/cls_fifo_legacy.cc:
```diff
@@ -404,7 +404,7 @@ std::optional<marker> FIFO::to_marker(std::string_view s)
 
   auto pos = s.find(':');
-  if (pos == string::npos) {
+  if (pos == s.npos) {
```
the 'before' is the idiomatic way, AFAIK.
(see for example https://en.cppreference.com/w/cpp/string/basic_string/npos)

tchaikov replied:
@ronen-fr i'd like to understand why `string::npos` is better than `s.npos` apart from that cppreference uses the former. i admit that i cannot find enough proof to prove that it the idiomatic way or disprove it, cppreference is not enough to prove this, IMHO. i agree that, it's a good source as a human-readable version of C++ standards. but if you could help me understand 'before' is *better*, i'd be more than happy change it.

ronen-fr replied:
'npos' is static, and that's one good reason to use it as string::npos. It's not a property of the specific string.
Also - it's not just cppref. That's the way you usually find it: in Microsoft docs, or (by chance) in the following
book: "Programming and Problem Solving with C++", and (judging by a google search) - "everywhere".

Don't have time now to look for the relevant standard doc.

cbodley replied:
one advantage of `s.npos` is that it's generic to the underlying string type - which, in this case, is a `std::string_view`, so `string::npos` wasn't actually correct

ronen-fr reviewed (commented):
Finished w/ the first 260 files.

idryomov commented:
rbd changes look good.

Just curious what protocol/strategy you followed, as in some cases you added e.g. `using std::cout` for just a couple of occurrences, in other cases you went for fixing each of about a dozen occurrences individually and in some cases just a blank `using namespace std` was added.

idryomov commented:
There is a typo in all commit messages: reply -> rely.

adamemerson reviewed (approved):
This is wonderful, thank you.

tchaikov commented:
changelog

- rebased onto master HEAD
- s/reply/rely/ in commit message

tchaikov commented:
> rbd changes look good.
> 
> Just curious what protocol/strategy you followed, as in some cases you added e.g. `using std::cout` for just a couple of occurrences, in other cases you went for fixing each of about a dozen occurrences individually and in some cases just a blank `using namespace std` was added.

the goals are (in the order of priority, from higher to lower)
1. to minimize the code churn, 
2. to preserve the existing coding style
3. to improve the readability
4. to reduce the usage of `using namespace std`.

trying to explain the the strategy i was following to archive the goals above briefly:

1. if there were already a bunch of declarations like `using std::string` in the beginning of the .cc file, i will try to preserve them by adding new `using std::*` to fulfill the needs
2. if the .cc source file only had a handful of, for instance, the `cout` references, instead of importing the whole `std` using `using namespace std`, i just added `std::` prefix to the type names. to reduce the usage of `using namespace std`.`
3. if there were many cases of unqualified C++ standard library names in .cc file, and they were not declared using `using ...`, despite that there were existing `using std::*` declarations, i just went ahead by replacing them with `using namespace std`, to minimize the code churn, instead of preserving the existing coding style.
4. in some cases where a `for` loop is involved, instead of replacing `map<foo, vector<pair<bar, baz>>>::iterator` with `std::map<foo, std::vector<pair<bar, baz>>>::iterator`, i just replaced it with `auto`, for better readability.
5. in some other cases, where "sv" or "s" literal postfix was used. to use them, we have to introduce the literal declarations into current naming space, but if "sv" or "s" postfix is not necessary and does not help with the readability, i just dropped them for better readability. as i think "less is more" in this case, as we, unlike C++ compiler frontend, take much more time when parsing more source code.

tchaikov commented:
Ronen Friedman ***@***.***>于2021年8月12日 周四13:36写道：

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/rgw/cls_fifo_legacy.cc
> <https://github.com/ceph/ceph/pull/42742#discussion_r687401690>:
>
> > @@ -404,7 +404,7 @@ std::optional<marker> FIFO::to_marker(std::string_view s)
>    }
>
>    auto pos = s.find(':');
> -  if (pos == string::npos) {
> +  if (pos == s.npos) {
>
> 'npos' is static, and that's one good reason to use it as string::npos.
> It's not a property of the specific string.
>

A static member variable *is* a member variable. So I am not convinced as I
am not able to see the connection.

>
> Also - it's not just cppref. That's the way you usually find it: in
> Microsoft docs, or (by chance) in the following
> book: "Programming and Problem Solving with C++", and (judging by a google
> search) - "everywhere".
>
They cannot prove the “better”.

> Don't have time now to look for the relevant standard doc.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/42742#discussion_r687401690>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAONP3RZMZJFVLN5DNQ3HLT4NMW5ANCNFSM5B5NP4AQ>
> .
>
-- 
Regards
Kefu Chai

rzarzynski commented on src/test/librbd/test_ObjectMap.cc:
```diff
@@ -17,6 +17,7 @@
 #include <boost/accumulators/statistics/stats.hpp>
 #include <boost/accumulators/statistics/rolling_sum.hpp>
 
+
```
Is the extra line intentional?

tchaikov replied:
no, fixed and repushed.

rzarzynski commented on src/client/ObjecterWriteback.h:
```diff
@@ -48,15 +48,13 @@ class ObjecterWriteback : public WritebackHandler {
-	 ++p)
-      op.write(p->first, p->second, trunc_size, trunc_seq);
+    for (auto& [offset, bl] : io_vec)
+      op.write(offset, bl, trunc_size, trunc_seq);
```
It's far easier to read :+1: .

rzarzynski commented on src/test/ObjectMap/KeyValueDBMemory.h:
```diff
@@ -158,13 +158,11 @@ class KeyValueDBMemory : public KeyValueDB {
-      string prefix = p->first.first;
-      bufferlist &bl = p->second;
+    for (auto& [key, bl] : db) {
+      string prefix = key.first;
```
Fine with that.

rzarzynski commented on src/cls/queue/cls_queue_client.h:
```diff
@@ -6,11 +6,11 @@
 
-#endif
\ No newline at end of file
+#endif
```
:+1:

rzarzynski reviewed (approved):
LGTM! No blocker detected.

rzarzynski commented on src/test/librbd/test_ObjectMap.cc:
```diff
@@ -158,6 +158,8 @@ TEST_F(TestObjectMap, AcquireLockInvalidatesWhenTooSmall) {
   ASSERT_TRUE(flags_set);
```
ACK.

idryomov commented on src/librbd/io/QosImageDispatch.cc:
```diff
@@ -56,10 +56,10 @@ QosImageDispatch<I>::QosImageDispatch(I* image_ctx)
-    m_throttles.push_back(make_pair(
-      flag.first,
-      new TokenBucketThrottle(cct, flag.second, 0, 0, timer, timer_lock)));
+  for (auto [flag, name] : throttle_flags) {
```
On that note, should that be changed to `for (const auto& [flag, name] : throttle_flags) {` to avoid unnecessary string copies?

rzarzynski replied:
That would be an optimization. However, I don't know whether we really should worry about this particular place (usually not). A profiler could answer.

tchaikov replied:
@idryomov i will create a follow-up PR to address your comment. this indeed came to my mind when making this change, but as @rzarzynski put, that'd be an optimization.  and i think this changes the behavior. so this is not part of this change.

tchaikov replied:
#42792

tchaikov commented:
https://pulpito.ceph.com/kchai-2021-08-13_12:30:46-rados-wip-kefu-testing-2021-08-13-1658-distro-basic-smithi/

wjwithagen commented:
@tchaikov 
Impressive patch set...
As far as I can see 1 omission for FreeBSD,. Is prefixing with `std::` the prefered solution??

```
/home/jenkins/workspace/ceph-master-compile/src/msg/async/EventKqueue.h:63:18: error: no template named 'vector'; did you mean 'std::vector'?
  int event_wait(vector<FiredFileEvent> &fired_events,
                 ^~~~~~
                 std::vector
/usr/include/c++/v1/vector:471:28: note: 'std::vector' declared here
class _LIBCPP_TEMPLATE_VIS vector
                           ^
/home/jenkins/workspace/ceph-master-compile/src/msg/async/EventKqueue.cc:216:30: error: no template named 'vector'; did you mean 'std::vector'?
int KqueueDriver::event_wait(vector<FiredFileEvent> &fired_events, struct timeval *tvp)
                             ^~~~~~
                             std::vector
/usr/include/c++/v1/vector:471:28: note: 'std::vector' declared here
class _LIBCPP_TEMPLATE_VIS vector
                           ^
2 errors generated.
```

tchaikov commented:
@wjwithagen yeah. i think it's preferred in the header files. and sorry for the FTBFS, could you help test #42776?
